### ISSUE TITLE
Attribute merge cache is not reset 

### ISSUE THREAD

**Issue by sun77** [chef/chef#4207]
Hello, 

This issue is linked to #3982

The following bug occurs with both chef 12.4.1 and 12.5.1.

We succeed to reproduce the issue in chef-shell : 
#### How to reproduce

``` ruby
chef:attributes > default['v1'] = 1
 => 1 
chef:attributes (12.5.1)> node['a']                                #(1)
 => nil 
chef:attributes (12.5.1)> default['a'] = node['v1']                #(2)
 => 1 
chef:attributes (12.5.1)> node['a']                                #(3)
 => nil 
```
1. When we access to the attribute node['a'], the merge process set the value to 'nil'.
2. When we set a value to node['a'] with another attribute, the merge process run on this attribute and set the "root.top_level_breadcrumb" value with this attribute.
   Here => "v1".
   So, the cache is reset on "v1" and not on "a".
3. Then, when we access to node['a'], we can see the value is not updated.
#### Conversely

Here, we access first to the node['v2'] attribute.

``` ruby
chef:attributes (12.5.1)> default['v2'] = 2
 => 2
chef:attributes (12.5.1)> node['v2']                               #(1)
 => 2 
chef:attributes (12.5.1)> node['b']                                #(2)
 => nil 
chef:attributes (12.5.1)> default['b'] = node['v2']                #(3)
 => 2 
chef:attributes (12.5.1)> node['b']                                #(4)
 => 2
```
1. We access first to the node['v2'] value.
2. When we access to the attribute node['b'], the merge process set the value to 'nil'.
3. When we set a value to node['b'] with another attribute, the merge process run correctly and reset the attribute "b"
4. Then, when we access to node['b'], we can see the value is updated because of the access in step 2.
## Conclusion

So, we strongly suppose there is a bug in the way to reset the attribute merge cache.

cc:@aboten

**Comment by Annih** [comment]
Hello,

We also found another kind of bug related to the implementation with `top_level_breadcrumb` in chef 12.

### How to reproduce
```ruby
chef:attributes (12.5.1)> a = default['a']                       #1
 => {}
chef:attributes (12.5.1)> node['a']                              #2
 => {}
chef:attributes (12.5.1)> default['b'] = 0                       #3
 => 0
chef:attributes (12.5.1)> a['key'] = 1                           #4
 => 1
chef:attributes (12.5.1)> node['a']                              #5
 => {}
```
1. Store a reference to `default['a']` (an empty mash) in a temporary variable
2. Compile/merge attribute `a`, resulting value is `{}`
3. Initialize a new attribute `b`, changing the `top_level_breadcrumb` to `b`
4. Set a value in the stored reference (default['a']), the reset mechanism try to reset `b`
5. Compiled/merged value of attribute 'a' is still `{}`

### Conclusion
This example is a bit more twisted because we store a reference to the `default` attribute mash, but it's sometime a really convenient shortcut :)
I think we should maybe keep a reference to the `top_level_breadcrumb` in each `VividMash` rathen than in the root.

**Comment by lamont-granquist** [comment]
So I think the second issue is easier to answer.  That is the crippling flaw of using global state (the top level breadcrumb) to track information.  So that's a good reason for why the current attributes implementation needs to be scrapped and rewritten.  I'll try to write up a spec test with that in it and apply it to the attributes rewrite branch and see if I've simply fixed that bug via the rewrite.  Offhand, though, I don't think I see a way to fix that, short of the rewrite or introducing some kind of awful hash of breadcrumbs indexed by precedence level and key and I really don't want to go that latter route since it just adds complexity and more bolted on state to have be buggy.

For now, I think the current implementation of attributes in Chef 12.x just does not support that kind of access where you store a node to a variable and expect to be able to fiddle with the node state and then go back to the variable and have it be sane.

**Comment by lamont-granquist** [comment]
Oh and I think I see the first problem as well...

We follow method chaining and construct the breadcrumb down the right hand side of the expression, but then we need to call `[]=(key, value)` so we must evaluate the `value` expr which is the `node['v1']` which clobbers the breadcrumb state.  I was thinking that maybe we could do something with the ordering inside of `#[]=` as to when we reset the cache, but the problem is that we only know that we've hit `#[]=` by the time its really too late and the breadcrumbs are all gone.

We also can't use the tactic of using a decorator to convert `node.default['foo']['bar']['baz'] =` into a functional `node.default_set('foo', 'bar', 'baz', value)` call because all those intermediate results have to be valid hash-like things (I do this in the v1.99 rewrite branch for stuff like `node.set_unless` because we don't consider the intermediate values for set_unless to be valid).

So I think you have to return intermediate decorators that carry their 'path' state along with them, and that's the whole purpose of the v1.99 rewrite...

I guess that's about all I've got.  Thanks for the very useful examples, at least now I've got a decent replication case rather than just issues that I couldn't seem to recreate, and got more incentive to get the rewrite shipped.

**Comment by lamont-granquist** [comment]
7f8bfe680f385c46780d8961825ff53f7fd36503 correct adds tests to my branch for both of these cases which fail on master and pass on my rewrite branch.

**Comment by christinedraper** [comment]
I suspect this is causing a problem I am seeing with deep merge and attributes, as well as some problems with chef provisioning. Appreciate if you could tell me if I am wrong. My scenario:

 With Mashes (as expected, mash1 is unaltered):
```
mash1 = Mash.new({"level1"=>{"level2"=>{"attr1"=>"val1"}}})
Chef::Mixin::DeepMerge.merge(mash1, {"level1"=>{"level2"=>{"attr1"=>"val2"}}})
mash1
{"level1"=>{"level2"=>{"attr1"=>"val1"}}}
```

With Attributes (broken, attributes are changed):
```
node1 = Chef::Node.new
node1.normal = {"level1"=>{"level2"=>{"attr1"=>"val1"}}}
Chef::Mixin::DeepMerge.merge(node1.normal, {"level1"=>{"level2"=>{"attr1"=>"val2"}}})
node1.normal
{"level1"=>{"level2"=>{"attr1"=>"val2"}}}
```

**Comment by robevansuk** [comment]
I found referencing a node in the following way doesn't work in Chef 12.6.0 (but used to in chef 11.16.4):

From an attributes file:
`default_service_name = default['service_name']
config = default_service_name['config']
config['var1'] = 'somevalue1'
config['var2'] = 'somevalue2'`

I've changed the config line to use a slightly more qualified path rather than using the reference to the reference:

`default_service_name = default[service_name]
config = default[service_name]['config']
config['var1'] = 'somevalue1'
config['var2'] = 'somevalue2'`

This resolves the issue for me but still doesn't really provide much in the way of explanation as to what's going on behind the scenes.

was able to test locally on a chef-client provisioned vagrant machine, however the behaviour is different between `vagrant provision` vs. logging into the box and running `sudo chef-client`. Seems vagrant provision does things the way we continued to expect them to whilst, chef-client does not.

Thought I'd share for what its worth.

**Comment by lamont-granquist** [comment]
yeah when you try to save the intermediate hash and then call it you clobber the breadcrumb state.  that is another manifestation of precisely the same bug

**Comment by lamont-granquist** [comment]
@christinedraper pretty sure that issue isn't this bug since you're looking at node.normal which is the accessor to the VividMash and you're not going in through `node["level1"]` which would produce a deep-merged and cached ImmutableMash result.

**Comment by phene** [comment]
Just ran into this after an hour of debugging.

**Comment by lamont-granquist** [comment]
can you test with https://github.com/chef/chef/pull/5378 and see if it fixes the issue?

based on what you wrote it doesn't sound like this issue unless you've omitted some other detail.

**Comment by lamont-granquist** [comment]
fixed by #5378 

```
chef:attributes (12.18.4)> a = default['a']
 => {}
chef:attributes (12.18.4)> node['a']
 => {}
chef:attributes (12.18.4)> default['b'] = 0
 => 0
chef:attributes (12.18.4)> a['key'] = 1
 => 1
chef:attributes (12.18.4)> node['a']
 => {"key"=>1}
```

if there's other issues on this thread that weren't fixed by that change, then those issues are different bugs, please open a new ticket.
